File: platform-unity/Core/Systems/DiagnosticsSystem.cs
using Browline.Hud.Core.Data;
using Browline.Hud.Core.Domain.Components;
using UnityEngine;

namespace Browline.Hud.Core.Systems;

public sealed class DiagnosticsSystem
{
    private const float MaxCardinalIndex = 7f;

    public void Update(DeviceDiagnosticsSnapshot snapshot, PowerState powerState, float deltaTimeMs)
    {
        var powerModifier = powerState == PowerState.On ? 1f : -1f;

        snapshot.CpuLoad = Clamp(snapshot.CpuLoad + Noise(deltaTimeMs, 0.08f) + powerModifier * 0.02f, 8f, 96f);
        snapshot.MemoryUsage = Clamp(snapshot.MemoryUsage + Noise(deltaTimeMs, 0.03f) + powerModifier * 0.01f, 6f, snapshot.MemoryCapacity - 2f);
        snapshot.NetworkStrength = Clamp(snapshot.NetworkStrength + Noise(deltaTimeMs, 0.002f), 0f, 1f);
        snapshot.NetworkConnection = snapshot.NetworkStrength > 0.2f ? NetworkConnectionType.Cellular5G : NetworkConnectionType.None;
        snapshot.TemperatureC = Clamp(snapshot.TemperatureC + Noise(deltaTimeMs, 0.001f) + powerModifier * 0.005f, 18f, 36f);
        snapshot.WindSpeedKmh = Clamp(snapshot.WindSpeedKmh + Noise(deltaTimeMs, 0.01f), 2f, 24f);

        if (Random.value > 0.98f)
        {
            var nextDirection = ((int)snapshot.WindDirection + 1) % ((int)MaxCardinalIndex + 1);
            snapshot.WindDirection = (CardinalDirection)nextDirection;
        }

        snapshot.HumidityPercent = Clamp(snapshot.HumidityPercent + Noise(deltaTimeMs, 0.005f), 20f, 70f);
    }

    private static float Noise(float deltaTimeMs, float amplitude)
    {
        var normalized = deltaTimeMs / 16.67f;
        return (Random.value - 0.5f) * 2f * amplitude * normalized;
    }

    private static float Clamp(float value, float min, float max)
    {
        if (value < min)
        {
            return min;
        }

        return value > max ? max : value;
    }
}

File: core/domain/systems/DiagnosticsSystem.ts
import { DiagnosticsComponent } from '../components/DiagnosticsComponent';
import { PowerState } from '../components/PowerComponent';
import { CardinalDirection, NetworkConnectionType } from '../../data/DeviceSpecifications';

const MAX_CARDINAL_INDEX = 7;

export class DiagnosticsSystem {
  public update(entityId: number, deltaTimeMs: number, powerState: PowerState): void {
    const powerModifier = powerState === PowerState.On ? 1 : -1;

    DiagnosticsComponent.cpuLoad[entityId] = this.clamp(
      DiagnosticsComponent.cpuLoad[entityId] + this.noise(deltaTimeMs, 0.08) + powerModifier * 0.02,
      8,
      96,
    );

    DiagnosticsComponent.memoryUsage[entityId] = this.clamp(
      DiagnosticsComponent.memoryUsage[entityId] + this.noise(deltaTimeMs, 0.03) + powerModifier * 0.01,
      6,
      DiagnosticsComponent.memoryCapacity[entityId] - 2,
    );

    DiagnosticsComponent.networkStrength[entityId] = this.clamp(
      DiagnosticsComponent.networkStrength[entityId] + this.noise(deltaTimeMs, 0.002),
      0,
      1,
    );

    DiagnosticsComponent.networkConnection[entityId] =
      DiagnosticsComponent.networkStrength[entityId] > 0.2
        ? NetworkConnectionType.Cellular5G
        : NetworkConnectionType.None;

    DiagnosticsComponent.temperatureC[entityId] = this.clamp(
      DiagnosticsComponent.temperatureC[entityId] + this.noise(deltaTimeMs, 0.001) + powerModifier * 0.005,
      18,
      36,
    );

    DiagnosticsComponent.windSpeedKmh[entityId] = this.clamp(
      DiagnosticsComponent.windSpeedKmh[entityId] + this.noise(deltaTimeMs, 0.01),
      2,
      24,
    );

    const shouldShiftDirection = Math.random() > 0.98;
    if (shouldShiftDirection) {
      const currentDirection = DiagnosticsComponent.windDirection[entityId];
      const nextDirection = (Math.round(currentDirection) + 1) % (MAX_CARDINAL_INDEX + 1);
      DiagnosticsComponent.windDirection[entityId] = nextDirection as CardinalDirection;
    }

    DiagnosticsComponent.humidityPercent[entityId] = this.clamp(
      DiagnosticsComponent.humidityPercent[entityId] + this.noise(deltaTimeMs, 0.005),
      20,
      70,
    );
  }

  private noise(deltaTimeMs: number, amplitude: number): number {
    const normalized = deltaTimeMs / 16.67;
    return (Math.random() - 0.5) * 2 * amplitude * normalized;
  }

  private clamp(value: number, min: number, max: number): number {
    return Math.min(Math.max(value, min), max);
  }
}
